quickwit/quickwit-janitor/src/retention_policy_execution.rs

```diff
@@ -41,22 +41,21 @@ pub async fn run_execute_retention_policy(
     retention_policy: &RetentionPolicy,
     ctx: &ActorContext<RetentionPolicyExecutor>,
 ) -> anyhow::Result<Vec<SplitMetadata>> {
-    // Select published splits and filter for expiration.
-    let current_date_time = OffsetDateTime::now_utc();
-
-    let query = ListSplitsQuery::for_index(index_id).with_split_state(SplitState::Published);
+    // Select splits that are published and older than the retention period.
+    let retention_period = retention_policy.retention_period()?;
+    let current_timestamp = OffsetDateTime::now_utc().unix_timestamp();
+    let max_retention_timestamp = current_timestamp - retention_period.as_secs() as i64;
+    let query = ListSplitsQuery::for_index(index_id)
+        .with_split_state(SplitState::Published)
+        .with_time_range_end_lte(max_retention_timestamp);
 
     let expired_splits: Vec<SplitMetadata> = ctx
         .protect_future(metastore.list_splits(query))
         .await?
         .into_iter()
-        // TODO: possibly move this in DB query
-        .filter(|split| {
-            is_split_expired(current_date_time, split, retention_policy).unwrap_or(false)
-        })
-        .map(|meta| meta.split_metadata)
+        .map(|split| split.split_metadata)
+        .filter(is_split_expired)
```

Comment:
Can we filter out splits without time range separately and emit a single warning?

Reply:
We had that but given that it needed a separate vector, @fmassot and I found this way better and concluded it was ok to warn on each split.